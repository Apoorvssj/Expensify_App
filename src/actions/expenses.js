import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';

//changed in v2 f15

//------see expenses-action-before-async.js in playground , to see before asynchronous action generator
//using startAddExpense as the function for async action(middle thunk needed,to dispatch functions)

//now in every other file where we dispatched addExpense, we will change that to startAddExpense , in AddExpensePage.js


// ADD_EXPENSE action generator
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//alternate way of setting up defualt values,rather than in function argument as done in old file in playground

export const startAddExpense = (expenseData = {}) => {
    //retuning a function not object,
    //this function is internally called by redux with dispatch
    return (dispatch) => {
        //destructuring expenseData , to set defualts
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        //es6 shortcut syntax to define properties in objects
        const expense = {description, note, amount, createdAt};
        //retuning for promise chaining in expense.test.js in actions
        return database.ref('expenses').push(expense).then((ref) => {
            //to change store
            dispatch(addExpense({
                id: ref.key, //now id is generated by firebase so need of uuid
                ...expense
            }));
        });
    };
};

//REMOVE_EXPENSE action generator
//default value is not needed for id, id : id using shorthand
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id 
    
});

//EDIT_EXPENSE action generator
//no need to set defualts,becoz if dont have the id we wont be updating it in the first place
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});