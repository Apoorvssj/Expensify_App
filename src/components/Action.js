import React from 'react';

//using shothand return syntax explained in OptionModal
const Action = (props) => (
      <div>
      <button 
      className="big-button"
      onClick={props.handlePick} 
      disabled={!props.hasOptions}>
      What should I do?
      </button>
      </div>
    );
  

  export default Action;