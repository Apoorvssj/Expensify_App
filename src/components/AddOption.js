import React from 'react';

export default class AddOption extends React.Component {
  state = {
      error: undefined
  };
    handleAddOption = (e) => {
      e.preventDefault();
      //passing e means grabbing the event
      const value = e.target.elements.option.value.trim();
      const error = this.props.handleAddOption(value);
    
       this.setState(() => ({error}));//same as error: error ,it is object shorthand in es6
  
       if(!error) {
         //if no error message the form clears out
         e.target.elements.option.value = '';
       }
    };
  
    //{this.state.error && <p>{this.state.error}</p>} here if this.state.error is undefined (which is falsy) <p> will not render ,if it has a value (it is truthy) <p> will be rendered
    render(){
      return (
        <div>
        {this.state.error && <p className="add-option-error">{this.state.error}</p>}
        <form className="add-option" onSubmit={this.handleAddOption}>
        <input className="add-option__input" type="text" name="option" />
        <button className="button">Add option</button>
        </form>
        </div>
      );
    }
  }