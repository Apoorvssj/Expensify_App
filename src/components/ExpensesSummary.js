import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

//i didnot used destructuring ,but andrew did in v13 f13

export const ExpensesSummary = (props) => {
    const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(props.expensesTotal / 100).format('$0,0.00');
    return (
        <div>
        {
            <h1>Viewing {props.expenseCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
        }
        </div>
    );

};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);