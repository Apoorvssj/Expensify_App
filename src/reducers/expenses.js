//Expenses Reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
      case 'ADD_EXPENSE':
          return [
              //spread operator
              ...state,
              action.expense
          ];
        case 'REMOVE_EXPENSE':
            return state.filter(( {id} ) => id !== action.id); //destructuring state array and taking id from it as id
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        //first adding all properties of expense array then overriding it with new values with spread operator
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense; //no chnage
                }
            });  
        case 'SET_EXPENSES':
            return action.expenses  ;   
      default:
          return state;
  }
};

export default expensesReducer;