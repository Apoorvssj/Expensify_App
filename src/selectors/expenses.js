//startdate and enddate filters changed to moment v13 f11

import moment from 'moment';

// Get visible expenses(filtering expenses)
//destructuring filters
const getVisibleExpenses = (expenses,{ text,sortBy,startDate,endDate }) => {
    //filter and sort both returns an array so we can chain them together to get filtered and sorted array
return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;//it will be always true if there is no startDate,becoz we will never filter things based off of that ,if it is not their(as we have given user ability to clear dates in react-dates in dashboard page)
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch; //if all of these are true the filter will return true and the array(expense) will be kept in expenses array
}).sort((a,b) => {
    if(sortBy === 'date'){
        return a.createdAt < b.createdAt ? 1 : -1; //if this is true 1 will be returned means b will come first,otherwise -1 will be returned means a will come first
    } 
    else if(sortBy === 'amount'){
        return a.amount < b.amount ? 1 : -1;
    }
});
};

export default getVisibleExpenses;

// before moment

/* 

{
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate; // always true for non numbers ,also for undefined .but dont worry we will let undefined filter it
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch; //if all of these are true the filter will return true and the array(expense) will be kept in expenses array
}

*/