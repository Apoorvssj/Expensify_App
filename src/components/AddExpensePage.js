import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

//v14 f12
//changing to class based component to avoid defining onSubmit function inline(which gets recalculated and redefined on every render),instead we will reference it

export class AddExpensePage extends React.Component {
   onSubmit = (expense) => {
// props.dispatch(addExpense(expense));//added all the data to the redux store
this.props.addExpense(expense);//adding this in order to test
this.props.history.push('/');//one of many props provided by router(can see components dev tools),without refresh it will got to dashboard page using browser routing,it is exactly equal to clicking a link
   };
   render() {
      return (
         <div>
         <h1>Add Expense</h1>
         <ExpenseForm 
           onSubmit={this.onSubmit}/>
      </div>
      );
   }
}
  

//export default AddExpensePage; without connect

//added this in order to test, as addExpense was an imported function,v14 f12,it is a way to return your dispatcher functions allowing you to abstract them away from the component,seee react-redux docs
const mapDisptachToProps = (dispatch) => ({
   addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDisptachToProps)(AddExpensePage); //first argument in mapStatesToProp,we dont need it so undefined