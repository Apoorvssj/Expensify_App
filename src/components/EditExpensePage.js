import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
// import {editExpense, removeExpense} from '../actions/expenses';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';

//refactored like AddExpensePage.js in v15 f12


//props.match.params.id is seen from console //v8 f9
export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
      this.props.startEditExpense(this.props.expense.id,expense);
      this.props.history.push('/');
    };
    onRemove = () => {
      this.props.startRemoveExpense({ id: this.props.expense.id });
      this.props.history.push('/');
    };
    render() {
       return (
         <div>
           <div className="page-header">
             <div className="content-container">
               <h1 className="page-header__title">Edit Expense</h1>
             </div>
           </div>
           <div className="content-container">
             <ExpenseForm
               expense={this.props.expense}
               onSubmit={this.onSubmit}
             />
             <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
           </div>
         </div>
       );
    }
};


const mapSateToProps = (state,props) => {
   return {
      expense: state.expenses.find((expense) => expense.id === props.match.params.id)
   };
};   

const mapDisptachToProps = (dispatch, props) => ({
   startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
      startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
   
  
});

// export default EditExpensePage;
export default connect(mapSateToProps, mapDisptachToProps)(EditExpensePage);