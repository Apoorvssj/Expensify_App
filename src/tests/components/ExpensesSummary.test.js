import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrappper = shallow(<ExpensesSummary expenseCount={1}
        expensesTotal={235}/>);
    expect(wrappper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
    const wrappper = shallow(<ExpensesSummary expenseCount={2}
        expensesTotal={435}/>);
    expect(wrappper).toMatchSnapshot();
});
