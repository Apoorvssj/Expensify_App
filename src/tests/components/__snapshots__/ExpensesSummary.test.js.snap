// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling 
    $4.35
  </h1>
</div>
`;
